### PR TITLE
Performance: Bootstrap metrics tracking

### PR DESCRIPTION
**Pull request by youknowriad**
### Changes proposed in this Pull Request:

**Big Picture** My goal is to bring metrics tracking and performance dashboard to the tools WooCommerce use daily to avoid performance regressions and work on performance improvements similarly to the [Gutenberg project](https://www.codevitals.run/project/gutenberg) for instance.

Implementing this require three complementary steps:

 1- First step is to have some test runner (e2e test here) that we can launch to compute the metrics.
 2- Second step is to be able to use the same test runner in two separate branches as part of the same CI job. The reason for this is that CI runners capacities are unstable and we can't rely on the absolute numbers. We have to compute relative numbers and compare each commit with a given "base commit".
 3- Third step is to send these values (base branch and commit) to the dashboard (codevitals) to log it and produce the graphs.

The current PR implements the first step. I'm not a WooCommerce developer, so I don't know which metrics are the most important to track for Woo but at least we should check Woo's impact on the editor (which I'm familiar with), so this PR adds a test that allows us to retrieve the editor's "typing" and "loading" metrics.

Woo developers should be able to use this as an example to add the metrics they want to track.

### How to test the changes in this Pull Request:

The test is only run locally at the moment, no CI changes. you can try it by setting up your "WooCommerce" plugin wp-env and then running:

```
cd plugins/woocommerce
pnpm test:metrics
```

you'll get something close to the following results:

Screenshot 2023-12-11 at 11 44 56 AM

---
I'm new to this repository, please let me know if I did anything wrong in terms of flows, code styles...

### CODE REVIEW

File: .syncpackrc [hunk 1]
```diff
@@ -171,6 +171,15 @@
...
 		},
+		{
+			"dependencies": [
+				"@wordpress/e2e-test-utils-playwright"
+			],
+			"packages": [
+				"**"
+			],
+			"pinVersion": "wp-6.4"
+		},
```
Comment (youknowriad) [inline .syncpackrc:182]: I was kind of forced to do that. I pinned this to the latest wp version (I saw this version used in woo blocks)

File: plugins/woocommerce/tests/metrics/config/global-setup.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import { request } from '@playwright/test';
```
Comment (youknowriad) [inline plugins/woocommerce/tests/metrics/config/global-setup.ts:1]: Most of these setup files are copied and adapted from the Gutenberg repository. There's probably an opportunity to share some of this through `@wordpress/scripts` but it seems out of scope.

File: plugins/woocommerce/tests/metrics/specs/editor.spec.js [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+				props.preferences
+			) ) {
+				await window.wp.data
+					.dispatch( 'core/preferences' )
+					.set( props.context, key, value );
+			}
+		},
+		{ context, preferences }
+	);
+}
```
Comment (youknowriad) [inline plugins/woocommerce/tests/metrics/specs/editor.spec.js:46]: These two utilities are actually present in the latest version of `@wordpress/e2e-test-utils-playwright`. We could just do `admin.editPost( id )` but I copied them over for now because of the pinned dependency version.

File: plugins/woocommerce/tests/metrics/assets/large-post.html
Comment (senadir) [inline plugins/woocommerce/tests/metrics/assets/large-post.html:1]: This is fine to track and compare to GB to make sure Woo isn't regressing GB but in general no Woo page or template would ever need this much data.

Comment (youknowriad) [reply plugins/woocommerce/tests/metrics/assets/large-post.html:1]: Yes, but regressions (especially for typing) are more visible in large posts.

Comment (nerrad) [reply plugins/woocommerce/tests/metrics/assets/large-post.html:1]: Agree with Riad here. We definitely, want to track how the registration of Woo Blocks adds to the overall editor performance. This will also be useful to see the impact of performance improvements made in GB (such as lazy block loading).

Comment (senadir) [reply plugins/woocommerce/tests/metrics/assets/large-post.html:1]: Ah yeah I agree, just saying that the basis overtime here won't be representative or very meaningful unless we're comparing the values against GB without Woo, so that we can see the real impact of Woo installed.

Comment (youknowriad) [reply plugins/woocommerce/tests/metrics/assets/large-post.html:1]: It's important to understand the numbers are not representative regardless of what we do. You shouldn't look at the numbers (example: 7seconds) and say, oh this is too much, regardless of metric (there are multiple factors that can skew the number, the CI load, the length of the post, the CPU...). The value is not meant to be a "realistic" value. The metric tracking value comes from the historical impact. it comes from comparing one commit to the previous one and seeing how that commit impacts the evolution of the metric. This ensure regressions don't slip when adding features.

**Comment by coreymckrill** [comment]
Reassigning my review to Solaris, as I think they're better suited to evaluate these changes.

**Review by senadir — APPROVED** [review]
I took this for a spin and it's showing up the results, this should be the first step into getting full metrics tracking in place.

**Review by alopezari — CHANGES_REQUESTED** [review]
Thank you very much for working on this @youknowriad!

This is a very interesting PR and it offers many benefits. We [already](https://github.com/woocommerce/woocommerce/tree/trunk/plugins/woocommerce/tests/performance) use [k6](https://k6.io/) for performance & load testing, which covers mostly HTTP requests so far but can be extended to cover [browser metrics](https://k6.io/docs/using-k6-browser/browser-metrics/) such as core webvitals and others.

I wonder if we should maybe extend the performance testing infrastructure rather than adding more E2E tests for this purpose as E2E should be focused more on product features and acceptance testing? I'm interested in your thoughts here.

Cc @tammullen 

However, I wanted to thank you again for the work you did here. Regardless of the method, it's great that we find ways to monitor and improve WooCommerce's performance 🔥 

Ps: I sent this comment as "request changes" just to make sure we confirm this is the appropriate way to continue with this approach before merging.

**Comment by nerrad** [comment]
@alopezari - this PR is very narrow in focus on Editor performance and is needed to ensure WooCommerce Blocks work doesn't impact that performance over time. With this in place, we'll be able to have a WooCommerce-specific view in the codevitals.run site ([example for Gutenberg](https://www.codevitals.run/project/gutenberg) that will help us track the impact of Woo Blocks related PRs and their impact on editor performance).

It is _crucial_ we have this in place sooner vs later because of the number of registered Woo Blocks, templates as well as the unique things we're doing (or will be doing) to extend the editor and block functionality.

Please don't delay this unnecessarily.

**Review by nerrad — APPROVED** [review]

**Comment by senadir** [comment]
Hey @alopezari 

When Riad and I initially worked on this PR, we looked into k6 tests, it was not clear if they run on each PR, or at each release, it's fine to include them in this. 

Also why E2E, we just need a way to load up an editor with Gutenberg and capture some browser provided metrics like when can someone interact with a page. If we can run playwright in the K6 suite, then I'm fine moving.

But like Darren said, I'd **love** if we can get the initial momentum going here, merge this low stake PR and its sister one, get at least 1 metric to show in codevitals (which is one Riad is helping us do), and then we can continue iterating on this on our own (move the code around, expand it, add stuff to it).

**Comment by senadir** [comment]
E2E is like I said a means to an end, and k6 can be considered a supplement to this.

The main goals we want to get as soon as possible are:
- See how Woo is impacting Gutenberg.
- Get a continuous history of said thing.

**Review by alopezari — APPROVED** [review]
Thank you all for the context! Seems I was missing important details. I'll approve this too 👍

**Comment by youknowriad** [comment]
Thanks :)  next step is going to be this PR https://github.com/woocommerce/woocommerce/pull/42693
